pkg/master/publish.go

```diff
@@ -147,11 +151,53 @@ func (m *Master) setEndpoints(serviceName string, ip net.IP, port int) error {
 			},
 		}
 	}
-	if !reflect.DeepEqual(e.Subsets, want) {
-		e.Subsets = want
-		glog.Infof("setting endpoints for master service %q to %v", serviceName, e)
+
+	// First, determine if the endpoint is in the format we expect (one
+	// subset, one port, N IP addresses).
+	formatCorrect, foundIP := m.checkEndpointSubsetFormat(e, ip.String(), port)
+	if !formatCorrect {
+		// Something is egregiously wrong, just re-make the endpoints record.
+		e.Subsets = []api.EndpointSubset{{
+			Addresses: []api.EndpointAddress{{IP: ip.String()}},
+			Ports:     []api.EndpointPort{{Port: port, Protocol: api.ProtocolTCP}},
+		}}
+		glog.Infof("Resetting endpoints for master service %q to %v", serviceName, e)
+		return m.endpointRegistry.UpdateEndpoints(ctx, e)
+	} else if !foundIP {
+		// So we *always* add our own IP address to the end, and if
+		// there are too many IP addresses, we remove from the
+		// beginning. Given the requirements stated at the top of this
+		// function, this should cause the list of IP addresses to
```

Comment:
Address structures are sorted by IP - what am I missing?  I guess it could NOT sort, but that would potentially trigger false updates.

Why not just detect that you have too many and remove one at random.  That should still converge, no?